### ISSUE TITLE
Destructuring with explicit type of nullable record is a parsing error.

### ISSUE THREAD

**Issue by water-mizuu** [dart-lang/sdk#52439]
As said in the title, the simplified code below is not accepted.

```dart
typedef Index = (int y, int x);

void main() {
  var (int y, int x) = (1, 3);
  var (bool reset, (int, int, int)? color) = switch ((y, x)) {
    Index ind when ind != ind => (true, (0, 128, 128)),
    _ => (false, null),
  };
}
```

Removing the question mark parses correctly, or by creating a type alias works as well.

```dart
typedef Index = (int y, int x);
typedef Color = (int r, int g, int b);

void main() {
  var (int y, int x) = (1, 3);
  var (bool reset, Color? color) = switch ((y, x)) {
    Index ind when ind != ind => (true, (0, 128, 128)),
    _ => (false, null),
  };
}
```

**Comment by johnniwinther** [comment]
cc @stereotype441 @jensjoha

**Comment by stereotype441** [comment]
Fixed in 284296f8628a56d22ec44f75a7220e6cbd5b466e.
